funds-manager/funds-manager-server/src/execution_client/swap.rs

```diff
@@ -0,0 +1,80 @@
+//! Handlers for executing swaps
+
+use std::sync::Arc;
+
+use ethers::{
+    providers::Middleware,
+    signers::LocalWallet,
+    types::{Address, Eip1559TransactionRequest, TransactionReceipt, U256},
+};
+use tracing::info;
+
+use crate::helpers::ERC20;
+
+use super::{error::ExecutionClientError, quotes::ExecutionQuote, ExecutionClient};
+
+impl ExecutionClient {
+    /// Execute a quoted swap
+    pub async fn execute_swap(
+        &self,
+        quote: ExecutionQuote,
+        wallet: LocalWallet,
+    ) -> Result<(), ExecutionClientError> {
+        // Approve the necessary ERC20 allowance
+        self.approve_erc20_allowance(
+            quote.sell_token_address,
+            quote.to,
+            quote.sell_amount,
+            &wallet,
+        )
+        .await?;
+
+        // Execute the swap
+        let receipt = self.execute_swap_tx(quote, wallet).await?;
```

Comment:
could we revoke the approved allowance if the swap execution fails? would prefer to leave no unused allowance on any external contracts

Reply:
you'll see in a follow up that is not possible as we separate out allowances from execution.